Feature/application cqrs

AlexBud369 opened the pull request:
#Commands and Queries

## Auth

### Added
- **RegisterCommand**  
  Handles user registration via `POST /auth/register` with JWT authentication

- **LoginCommand**  
  Handles user login via `POST /auth/login` with JWT authentication

## Users

### Implemented
- **BlockUserCommand**  
  Manages user blocking via `POST /users/{id}/block` (admin functionality)

- **ChangeRoleCommand**  
  Handles role updates via `POST /users/{id}/role` (admin functionality)

### Added
- **GetUserByIdQuery**  
  Retrieves user by ID via `GET /users/{id}`

- **GetAllUsersQuery**  
  Retrieves all users via `GET /users` with pagination and filtering

## Tags

### Added
- **CreateTagCommand**  
  Creates tags via `POST /tags` for book categorization

### Implemented
- **GetTagByIdQuery**  
  Retrieves tag by ID via `GET /tags/{id}`

- **GetAllTagsQuery**  
  Retrieves all tags via `GET /tags` with pagination

## Books

### Added
- **CreateBookCommand**  
  Manages book creation via `POST /books`

- **UpdateBookCommand**  
  Handles book updates via `PUT /books/{id}`

- **DeleteBookCommand**  
  Manages book deletion via `DELETE /books/{id}`

- **UploadBookCoverCommand**  
  Handles cover uploads via `POST /books/{id}/cover`

### Implemented
- **GetBookByIdQuery**  
  Retrieves book by ID via `GET /books/{id}`

- **GetAllBooksQuery**  
  Retrieves all books via `GET /books` with pagination, filtering and sorting

AlexBud369 committed:
feat: add book queries
- Implements GetAllBooksQuery with pagination and filtering
- Implements GetBookByIdQuery for retrieving a single book

AlexBud369 committed:
feat: add book creation command

AlexBud369 committed:
fix: add user existence validation in CreateBookCommand handler

AlexBud369 committed:
refactor: replace standard exception with custom BookNotFoundException and renamed file

AlexBud369 committed:
feat: add UpdateBook command with validation and tag handling
- Implemented book update logic with user ownership check
- Added validation for user, book existence and permissions
- Supported tag updates (clearing old tags and adding new ones)

AlexBud369 committed:
feat: add command for book removing

AlexBud369 committed:
feat: add UploadBookCoverCommand for updating book cover image

AlexBud369 committed:
fix: update DeleteBookCommandHandler to allow admin deletion
- Adds UserManager to check for Admin role
- Allows admins to delete any book
- Fixes DeleteBookCommandHandler to implement IRequestHandler

AlexBud369 committed:
feat: add BlockUserCommand to allow admins to block users

AlexBud369 committed:
feat: add ChangeRoleCommand to alow admins to change user roles

AlexBud369 committed:
feat: add GetAllUsersQuery for viewing all users

AlexBud369 committed:
feat: add a command to retrieve a user by ID

AlexBud369 committed:
feat: add command for user registration

AlexBud369 committed:
feat: add a command to log a user in system

AlexBud369 committed:
feat: add creation Tag command

AlexBud369 committed:
feat: add query for viewing all tags

AlexBud369 committed:
feat: add query to get tag by ID

AlexBud369 committed:
fix: update Book handlers
- fixes CreateBookCommandHandler method name and CreatedByUserId
- updates GetAllBooksQueryHandler to use AppDbContext
- updates exception message

AlexBud369 committed:
feat: add command to allow admins to delete users

AlexBud369 committed:
fix: enhances RegisterCommandHandler with email and username uniqueness checks

budnikme commented on src/Application/Features/Auth/Commands/LoginCommandHandler.cs:
```diff
@@ -0,0 +1,66 @@
+        var user = await _userManager.FindByEmailAsync(request.Email);
+        if (user == null)
+        {
+            throw new UserNotFoundException($"User with email {request.Email} was not found.");
```
Please create Guard/utils class that will handle all exceptions, such as null. empty string and etc. Also, Mmybe it would be worth in your case to create one general exception instead of multiple

budnikme commented on src/Application/Features/Auth/Commands/LoginCommandHandler.cs:
```diff
@@ -0,0 +1,66 @@
+            throw new UnauthorizedAccessException("Invalid password.");
+        }
+
+        var claims = new List<Claim>
```
Please move that to a dedicated method, like CreateUserClaims

budnikme commented on src/Application/Features/Auth/Commands/LoginCommandHandler.cs:
```diff
@@ -0,0 +1,66 @@
+        var roles = await _userManager.GetRolesAsync(user);
+        claims.AddRange(roles.Select(role => new Claim(ClaimTypes.Role, role)));
+
+        var key = new SymmetricSecurityKey(Encoding.UTF8.GetBytes(_configuration["Jwt:Key"]));
```
Please move that to a dediacted method also.

budnikme commented on src/Application/Features/Auth/Commands/LoginCommandHandler.cs:
```diff
@@ -0,0 +1,66 @@
+        claims.AddRange(roles.Select(role => new Claim(ClaimTypes.Role, role)));
+
+        var key = new SymmetricSecurityKey(Encoding.UTF8.GetBytes(_configuration["Jwt:Key"]));
+        var creds = new SigningCredentials(key, SecurityAlgorithms.HmacSha256);
```
Generally speaking implementing and maintaining authentication by ourselves is considered a bad practice. In production we almost always use ready solutions like .net identity, openiddict or auth0

budnikme commented on src/Application/Features/Books/Commands/CreateBookCommandHandler.cs:
```diff
@@ -0,0 +1,55 @@
+
+            var newTagNames = tagNames.Except(existingTags.Select(t => t.tagName)).ToList();
+
+            book.TagsList.AddRange(newTagNames.Select(name => new Tag { tagName = name }));
```
It would be better to move all business logic to services

budnikme commented on src/Application/Features/Books/Commands/UpdateBookCommandHandler.cs:
```diff
@@ -0,0 +1,58 @@
+        }
+
+        _mapper.Map(request.Book, book);
+        if (request.Book.Tags != null)
```
Please use this styling instead, it looks more simplier and better
```
if (request.Book.Tags == null) {
    return;
}
```

budnikme commented on src/Application/Features/Books/Commands/UploadBookCoverCommandHandler.cs:
```diff
@@ -0,0 +1,45 @@
+            throw new UnauthorizedAccessException("Only the book owner can upload a cover image");
+        }
+
+        book.CoverImageUrl = request.CoverImageUrl;
```
Please create a dedicated service for images/files processing

AlexBud369 committed:
feat: refactor LoginCommandHandler with IAuthService and localization
- Replaced custom exceptions (UserNotFoundException, UnauthorizedAccessException) with Guard and ApplicationException
- Added localization support with IStringLocalizer<SharedResource> and Guard.Initialize
- Added refresh token creation and storage using AppDbContext
- Updated AuthResponseDto to include RefreshToken field
- Extracted JWT and refresh token generation to IAuthService for better separation of concerns
- Removed Creation UserClaims and Creation JwtCredentials commands as they are now in AuthService

AlexBud369 committed:
feat: refactor RegisterCommandHandler with IAuthService and localization
- Replaced InvalidOperationException and UnauthorizedAccessException with Guard and ApplicationException
- Added localization support with IStringLocalizer<SharedResource> and Guard.Initialize
- Extracted user creation and role assignment to IAuthService for better separation of concerns
- Updated UserDto mapping to include Role via opts.Items

AlexBud369 committed:
feat: add IAuthService and AuthService for authentication logic
- Created IAuthService interface with methods for JWT token generation, refresh token generation, and user creation
- Implemented AuthService to handle authentication-related business logic
- Moved JWT and refresh token logic from LoginCommandHandler to AuthService
- Moved user creation and role assignment logic from RegisterCommandHandler to AuthService

AlexBud369 committed:
refactor: enhance CreateBookCommandHandler with Guard, localization, and ITagService
- Replaced UserNotFoundException with Guard.AgainstNull and ApplicationException
- Extracted tag processing logic to ITagService to address reviewer feedback on business logic separation
- Added localization support with IStringLocalizer<SharedResource> and Guard.Initialize

AlexBud369 committed:
refactor: enhance DeleteBookCommandHandler with Guard, localization, and IBookAccessService
- Replaced UserNotFoundException, BookNotFoundException, and UnauthorizedAccessException with Guard.AgainstNull and Guard.AgainstUnauthorized
- Extracted access validation logic to IBookAccessService to address reviewer feedback on business logic separation
- Added localization support with IStringLocalizer<SharedResource> and Guard.Initialize

AlexBud369 committed:
refactor: enhance UpdateBookCommandHandler with Guard, localization, ITagService, and IBookAccessService
- Replaced UserNotFoundException, BookNotFoundException, and UnauthorizedAccessException with Guard.AgainstNull and Guard.AgainstUnauthorized
- Extracted tag processing logic to ITagService to address reviewer feedback on business logic separation
- Extracted access validation logic to IBookAccessService
- Added localization support with IStringLocalizer<SharedResource> and Guard.Initialize

AlexBud369 committed:
refactor: Enhance UploadBookCoverCommandHandler
- Replaced UserNotFoundException, BookNotFoundException, and UnauthorizedAccessException with Guard.AgainstNull and Guard.AgainstUnauthorized
- Extracted access validation logic to IBookAccessService to address reviewer feedback on business logic separation
- Extracted cover image processing to IImageService
- Added localization support with IStringLocalizer<SharedResource> and Guard.Initialize

AlexBud369 committed:
refactor: update GetAllBooksQueryHandler to use IBookQueryService
- Replaced direct EF Core queries with IBookQueryService for pagination and filtering
- Removed IMapper dependency as mapping is handled in BookQueryService
- Retained Guard and localization via IBookQueryService

AlexBud369 committed:
refactor: update GetBookByIdQueryHandler to use IBookQueryService
- Replaced direct EF Core queries and BookNotFoundException with IBookQueryService
- Removed IMapper dependency as mapping is handled in BookQueryService
- Retained Guard and localization via IBookQueryService

AlexBud369 committed:
refactor: update GetAllTagsQueryHandler to use ITagQueryService
- Replaced direct EF Core queries with ITagQueryService for pagination and filtering
- Removed IMapper dependency as mapping is handled in TagQueryService
- Retained Guard and localization via ITagQueryService

AlexBud369 committed:
refactor: update GetTagByIdQueryHandler to use ITagQueryService
- Replaced direct EF Core queries and TagNotFoundException with ITagQueryService
- Removed IMapper dependency as mapping is handled in TagQueryService
- Retained Guard and localization via ITagQueryService

AlexBud369 committed:
refactor: update GetAllUsersQueryHandler to use IUserQueryService
- Replaced direct EF Core queries with IUserQueryService for pagination and filtering
- Removed IMapper dependency as mapping is handled in UserQueryService
- Retained Guard and localization via IUserQueryService

AlexBud369 committed:
refactor: update GetUserByIdQueryHandler to use IUserQueryService
- Replaced direct EF Core queries and UserNotFoundException with IUserQueryService
- Removed IMapper dependency as mapping is handled in UserQueryService
- Retained Guard and localization via IUserQueryService

AlexBud369 committed:
refactor: update CreateTagCommandHandler with Guard and localization
- Added IStringLocalizer for localized error messages
- Replaced UnauthorizedAccessException with Guard.Against for duplicate tag validation

AlexBud369 committed:
refactor: update BlockUserCommandHandler with Guard and localization
- Added IStringLocalizer for localized error messages
- Replaced UserNotFoundException and AdminOnlyAccessException with Guard.Against

AlexBud369 committed:
refactor: update ChangeRoleCommandHandler with Guard and localization
- Added IStringLocalizer for localized error messages
- Replaced UserNotFoundException, AdminOnlyAccessException, and UnauthorizedAccessException with Guard.Against

AlexBud369 committed:
refactor: update DeleteUserCommandHandler with Guard and localization
- Added IStringLocalizer for localized error messages
- Replaced UserNotFoundException and AdminOnlyAccessException with Guard.Against

AlexBud369 committed:
feat: add IImageService and ImageService
- Created IImageService interface for managing book cover uploads
- Implemented ImageService to coordinate uploads with IStorageService
- Integrated Guard and IStringLocalizer for validation and localization

AlexBud369 committed:
feat: add IBookAccessService and BookAccessService
- Created IBookAccessService interface for book access validation
- Implemented BookAccessService to check user permissions for book operations
- Integrated with UserManager for role checks and Guard for validation

AlexBud369 committed:
feat: add ITagService and TagService
- Created ITagService interface for tag management operations
- Implemented TagService to handle tag creation and validation
- Integrated Guard and IStringLocalizer for validation and localization

AlexBud369 committed:
feat: add ITagQueryService and TagQueryService
- Created ITagQueryService interface for tag retrieval operations
- Implemented TagQueryService to handle tag queries

AlexBud369 committed:
feat: add IUserQueryService and UserQueryService
- Created IUserQueryService interface for user retrieval operations
- Implemented UserQueryService to handle user queries for admin panel

AlexBud369 committed:
feat: add IFileStorageServic interface to abstract file storage operations

AlexBud369 committed:
feat: add LogoutCommand and Handler for user logout
- Created LogoutCommand to specify user ID and refresh token for logout
- Implemented LogoutCommandHandler to mark refresh token as revoked

AlexBud369 committed:
feat: add BookQueryService and IBookQueryService
- Implemented IBookQueryService with GetBooksAsync and GetBookByIdAsync methods
- Added BookQueryService with PagedResponseDto<BookDto> and TagIds filtering
- Included Guard checks for dependencies in constructor
- Replaced magic strings with DomainConstants
- Added localization for error messages via IStringLocalizer<SharedResource>

AlexBud369 committed:
refactor: update GetAllBooksQuery and GetAllBooksQueryHandler
- Refactored GetAllBooksQuery to use BookFilterDto
- Updated GetAllBooksQueryHandler to return PagedResponseDto<BookDto> instead of IEnumerable<BookDto>
- Added Guard checks for localizer, bookQueryService, and request.Filter
- Ensured all BookFilterDto parameters (including TagIds) are passed to GetBooksAsync

AlexBud369 commented:
### Core Architectural Improvements

#### Command Handlers Refactoring
- **Standardized exception handling**  
  Replaced custom exceptions with `Guard` pattern and `ApplicationException`  
  *(Ensures consistent validation across application)*

- **Implemented localization**  
  Added `IStringLocalizer<SharedResource>` integration  
  *(Enables multilingual support for error messages)*

- **Business logic separation**  
  Extracted core logic to dedicated service layer  
  *(Improves maintainability and testability)*

#### Query Handlers Optimization
- **Abstracted data access**  
  Replaced direct EF Core calls with specialized query services  
  *(Centralizes data retrieval logic)*

- **Maintained validation**  
  Preserved `Guard` pattern usage  
  *(Consistent authorization checks)*

### New Service Layer Components

#### Authentication Services
- `IAuthService` (interface) - Defines auth operations contract
- `AuthService` (implementation) - Handles JWT, refresh tokens, user management

#### Book Management Services
- `IBookAccessService` - Book permission validation logic
- `IBookQueryService` - Book data retrieval with pagination
- `BookQueryService` - Concrete implementation

#### Tag Processing Services
- `ITagService` - Tag CRUD operations
- `ITagQueryService` - Tag query operations
- `TagService`/`TagQueryService` - Implementations

#### User Management
- `IUserQueryService` - User data retrieval
- `UserQueryService` - Implementation

#### Media Handling
- `IImageService` - Book cover processing
- `IFileStorageService` - File storage abstraction

AlexBud369 committed:
feat: enhance IImageService and ImageService
- Added DeleteBookCoverAsync method to IImageService and ImageService
- Added optional deleteOldCover parameter to UpdateBookCoverAsync
- Renamed IStorageService to IFileStorageService for consistency
- Added unique file name generation in UpdateBookCoverAsync
- Added logic to delete old cover before uploading new one

AlexBud369 committed:
feat: update UploadBookCoverCommand and handler
- Changed UploadBookCoverCommand to use IFormFile instead of CoverImageUrl
- Updated handler to process IFormFile and pass stream to IImageService
- Moved file name generation to ImageService
- Added Microsoft.AspNetCore.Http for IFormFile support

AlexBud369 committed:
feat: add DeleteFileAsync method to support file deletion

budnikme commented on src/Application/Services/BookQueryService.cs:
```diff
@@ -0,0 +1,105 @@
+
+        query = sortBy switch
+        {
+            "title" => sortDescending
```
Please use enums instead of raw strings in such cases

budnikme commented on src/Application/Services/BookQueryService.cs:
```diff
@@ -0,0 +1,105 @@
+        Guard.Initialize(_localizer);
+    }
+
+    public async Task<PagedResult<BookDto>> GetBooksAsync(
```
This method is too big, please refactor it. Ideally, one method shouldn't have more than 5-10 rows

AlexBud369 committed:
fix: enhance Books query handlers with proper localization and validation
- Replaced IStringLocalizer<SharedResource> with IStringLocalizer<SharedResources> for consistency with other services
- Updated Guard.Against* to use localizer.GetString(SharedResources.<Key>) for localized error messages
- Added Guard.AgainstEmptyGuid in GetBookByIdQueryHandler to validate request.Id, ensuring robust input validation

AlexBud369 committed:
refactor: enhance BookQueryService for better readability and type safety
- Split monolithic GetBooksAsync into smaller methods  for improved readability and maintainability
- Replaced string-based sortBy with BookSortBy enum to eliminate magic strings and improve type safety
- Replaced IStringLocalizer<SharedResource> with IStringLocalizer<SharedResources> for consistency
- Updated Guard.Against* to use localizer.GetString(SharedResources.<Key>) for localized error messages

AlexBud369 committed:
fix: enhance Auth command handlers with proper localization and dependency validation
- Replaced IStringLocalizer<SharedResource> with IStringLocalizer<SharedResources> for consistency across the codebase.
- Updated Guard.Against* to use localizer.GetString(SharedResources.<Key>) for localized error messages
- Added dependency validation in constructors using Guard.AgainstNull for UserManager, IMapper, IConfiguration, AppDbContext, and IAuthService with localized messages.

AlexBud369 committed:
fix: enhance Books command handlers with proper localization and validation
- Replaced IStringLocalizer<SharedResource> with IStringLocalizer<SharedResources> for consistency
- Updated Guard.Against* to use localizer.GetString(SharedResources.<Key>) for localized error messages
- Added dependency validation in constructors for AppDbContext, IMapper, ITagService, UserManager, IBookAccessService, and IImageService with localized messages.
- Added Guard.AgainstEmptyGuid for Id and UserId validation across all handlers to ensure robust input validation.

AlexBud369 committed:
fix: enhance Tags command and query handlers with proper localization and validation
- Replaced IStringLocalizer<SharedResource> with IStringLocalizer<SharedResources> in CreateTagCommandHandler, GetAllTagsQueryHandler, and GetTagByIdQueryHandler for consistency.
- Updated Guard.Against* to use localizer.GetString(SharedResources.<Key>) for localized error messages
- Added dependency validation in constructors for AppDbContext, IMapper, and ITagQueryService with localized messages
- Added Guard.AgainstEmptyGuid for Id validation in GetTagByIdQueryHandler.
- Added Guard.AgainstInvalidPageNumber and Guard.AgainstInvalidPageSize in GetAllTagsQueryHandler for robust input validation

AlexBud369 committed:
fix: Enhance Users command and query handlers with proper localization and validation

AlexBud369 committed:
fix: Enhance services with proper localization and validation

AlexBud369 commented:
## Handler and Service Refactoring

### Book Handlers
- Standardized localization by replacing `IStringLocalizer<SharedResource>` with `IStringLocalizer<SharedResources>`
- Updated `Guard.Against*` checks to use localized messages via `localizer.GetString(SharedResources.<Key>)`
- Added `Guid.Empty` validation for `request.Id` in `GetBookByIdQueryHandler`
- Implemented dependency validation in constructors (`AppDbContext`, `IMapper`, `ITagService`, etc.)

### Auth Handlers
- Unified localization to use `IStringLocalizer<SharedResources>`
- Enhanced `Guard.Against*` checks with proper localization
- Added constructor dependency validation with localized messages

### Tag Handlers
- Updated localization in all tag handlers
- Added pagination validation (`Guard.AgainstInvalidPageNumber/Size`)
- Implemented `Guid.Empty` validation for IDs
- Added dependency validation for data access components

### User Handlers
- Standardized localization across user operations
- Added ID validation for all user-related commands
- Implemented pagination validation with domain constants
- Enhanced constructor validation

### Services
- **Global Changes**:
  - Unified localization to `IStringLocalizer<SharedResources>`
  - Improved `Guard.Against*` implementation
  - Added comprehensive constructor validation

- **Service-Specific Improvements**:
  - **BookQueryService**:
    - Split `GetBooksAsync` into smaller methods
    - Replaced string sorting with `BookSortBy` enum
  - **ImageService**:
    - Replaced hardcoded extensions with `DomainConstants.BookCover.AllowedImageExtensions`
  - **Tag Services**:
    - Added `TagName` validation using `DomainConstants.Tag.NameMaxLength`

budnikme commented on src/Application/Features/Auth/Commands/LoginCommandHandler.cs:
```diff
@@ -0,0 +1,76 @@
+        AppDbContext context,
+        IAuthService authService)
+    {
+        Guard.AgainstNull(userManager, nameof(userManager), localizer.GetString(SharedResources.UserManagerRequired));
```
.
